## **In general, consumer loans and credit cards are the most typical retail products where credit risk modelling is applied. In this project, we will focus on consumer loans.**

---

### In this project, we will use an open-source dataset titled "Lending Club Loan Data," which contains comprehensive information on all loans issued by Lending Club, a U.S. based peer-to-peer lending company from 2007 to 2015. The dataset is available at https://www.kaggle.com/datasets/adarshsng/lending-club-loan-data-csv?resource=download. 

---

### The separate files are provided,

#### 1) loan.csv : Complete dataset.
#### 2) LCDataDictionary.xlsx : Detailing the variables in the Dataset.

---
### Libraries used.

> !pip install pandas
> 
> !pip install numpy
> 
> !pip install openpyxl
> 
> !pip install seaborn
> 
> !pip install scikit-learn

In [1]:
import warnings
import pandas as pd
from IPython.display import display, Markdown

In [2]:
warnings.filterwarnings('ignore') # to supress warnings

> ### **First we will have a look at the dataset, the variables, and the description provided.**

In [3]:
variables = pd.read_excel('dataset/LCDataDictionary.xlsx')
loan_dataset = pd.read_csv('dataset/loan.csv') # Loading data.

> ### **After exploring the variables in the dataset, some stood out as important for our credit risk model. Their definitions are provided below.**

In [4]:
imp_variables = ['grade', 'emp_length', 'emp_title', 'term', 'earliest_cr_line', 'issue_d',
                'sub_grade', 'home_ownership', 'verification_status', 'loan_status', 'purpose',
                 'addr_state', 'initial_list_status','total_rev_hi_lim', 'annual_inc', 
                 'acc_now_delinq', 'total_acc', 'pub_rec', 'open_acc', 'inq_last_6mths',
                 'delinq_2yrs','dti', 'mths_since_last_record', 'mths_since_last_delinq', 
                'funded_amnt', 'loan_amnt', 'funded_amnt_inv']

for i in range(len(variables['LoanStatNew'])):
    if variables['LoanStatNew'][i] in imp_variables:
        display(Markdown(f"**{variables['LoanStatNew'][i]}**: {variables['Description'][i]}  \n"))

**acc_now_delinq**: The number of accounts on which the borrower is now delinquent.  


**addr_state**: The state provided by the borrower in the loan application  


**annual_inc**: The self-reported annual income provided by the borrower during registration.  


**delinq_2yrs**: The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years  


**dti**: A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.  


**earliest_cr_line**: The month the borrower's earliest reported credit line was opened  


**emp_length**: Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.   


**emp_title**: The job title supplied by the Borrower when applying for the loan.*  


**funded_amnt**: The total amount committed to that loan at that point in time.  


**funded_amnt_inv**: The total amount committed by investors for that loan at that point in time.  


**grade**: LC assigned loan grade  


**home_ownership**: The home ownership status provided by the borrower during registration or obtained from the credit report. Our values are: RENT, OWN, MORTGAGE, OTHER  


**initial_list_status**: The initial listing status of the loan. Possible values are – W, F  


**inq_last_6mths**: The number of inquiries in past 6 months (excluding auto and mortgage inquiries)  


**issue_d**: The month which the loan was funded  


**loan_amnt**: The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value.  


**loan_status**: Current status of the loan  


**mths_since_last_delinq**: The number of months since the borrower's last delinquency.  


**mths_since_last_record**: The number of months since the last public record.  


**open_acc**: The number of open credit lines in the borrower's credit file.  


**pub_rec**: Number of derogatory public records  


**purpose**: A category provided by the borrower for the loan request.   


**sub_grade**: LC assigned loan subgrade  


**term**: The number of payments on the loan. Values are in months and can be either 36 or 60.  


**total_acc**: The total number of credit lines currently in the borrower's credit file  


**verification_status**: Indicates if income was verified by LC, not verified, or if the income source was verified  


In [5]:
# To get pandas to show all columns in our dataframe, we use
pd.options.display.max_columns = None # To show all rows, we could use "pd.options.display.max_rows = None"
loan_dataset[:10]

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       2500         2500           2500.0   36 months   
1 NaN        NaN      30000        30000          30000.0   60 months   
2 NaN        NaN       5000         5000           5000.0   36 months   
3 NaN        NaN       4000         4000           4000.0   36 months   
4 NaN        NaN      30000        30000          30000.0   60 months   
5 NaN        NaN       5550         5550           5550.0   36 months   
6 NaN        NaN       2000         2000           2000.0   36 months   
7 NaN        NaN       6000         6000           6000.0   36 months   
8 NaN        NaN       5000         5000           5000.0   36 months   
9 NaN        NaN       6000         6000           6000.0   36 months   

   int_rate  installment grade sub_grade            emp_title emp_length  \
0     13.56        84.92     C        C1                 Chef  10+ years   
1     18.94       777.23     D        D2          Postmaster   10+ years   
2     17.97       180.69     D        D1       Administrative    6 years   
3     18.94       146.51     D        D2        IT Supervisor  10+ years   
4     16.14       731.78     C        C4             Mechanic  10+ years   
5     15.02       192.45     C        C3         Director COE  10+ years   
6     17.97        72.28     D        D1      Account Manager    4 years   
7     13.56       203.79     C        C1   Assistant Director  10+ years   
8     17.97       180.69     D        D1  Legal Assistant III  10+ years   
9     14.47       206.44     C        C2                  NaN   < 1 year   

  home_ownership  annual_inc verification_status   issue_d loan_status  \
0           RENT     55000.0        Not Verified  Dec-2018     Current   
1       MORTGAGE     90000.0     Source Verified  Dec-2018     Current   
2       MORTGAGE     59280.0     Source Verified  Dec-2018     Current   
3       MORTGAGE     92000.0     Source Verified  Dec-2018     Current   
4       MORTGAGE     57250.0        Not Verified  Dec-2018     Current   
5       MORTGAGE    152500.0        Not Verified  Dec-2018     Current   
6           RENT     51000.0     Source Verified  Dec-2018     Current   
7           RENT     65000.0     Source Verified  Dec-2018     Current   
8       MORTGAGE     53580.0     Source Verified  Dec-2018     Current   
9            OWN    300000.0        Not Verified  Dec-2018     Current   

  pymnt_plan  url desc             purpose                    title zip_code  \
0          n  NaN  NaN  debt_consolidation       Debt consolidation    109xx   
1          n  NaN  NaN  debt_consolidation       Debt consolidation    713xx   
2          n  NaN  NaN  debt_consolidation       Debt consolidation    490xx   
3          n  NaN  NaN  debt_consolidation       Debt consolidation    985xx   
4          n  NaN  NaN  debt_consolidation       Debt consolidation    212xx   
5          n  NaN  NaN         credit_card  Credit card refinancing    461xx   
6          n  NaN  NaN  debt_consolidation       Debt consolidation    606xx   
7          n  NaN  NaN         credit_card  Credit card refinancing    460xx   
8          n  NaN  NaN  debt_consolidation       Debt consolidation    327xx   
9          n  NaN  NaN  debt_consolidation       Debt consolidation    068xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  18.24          0.0         Apr-2001             1.0   
1         LA  26.52          0.0         Jun-1987             0.0   
2         MI  10.51          0.0         Apr-2011             0.0   
3         WA  16.74          0.0         Feb-2006             0.0   
4         MD  26.35          0.0         Dec-2000             0.0   
5         IN  37.94          0.0         Sep-2002             3.0   
6         IL   2.40          0.0         Nov-2004             1.0   
7         IN  30.10          0.0         Nov-1997             0.0   
8         FL  21.16          0.0         Aug-

> ### **Important thing to notice is that there are a lot of missing values or NAN. So, first we can view the columns that have all values missing.**

In [6]:
# This returns the toal number of null values present in each column.
pd.options.display.max_rows = None
loan_dataset.isnull().sum()

id                                            2260668
member_id                                     2260668
loan_amnt                                           0
funded_amnt                                         0
funded_amnt_inv                                     0
term                                                0
int_rate                                            0
installment                                         0
grade                                               0
sub_grade                                           0
emp_title                                      166969
emp_length                                     146907
home_ownership                                      0
annual_inc                                          4
verification_status                                 0
issue_d                                             0
loan_status                                         0
pymnt_plan                                          0
url                         

In [7]:
# This gives us the name of the columns where all entries are missing.
loan_dataset.columns[loan_dataset.isna().all()].tolist()

['id', 'member_id', 'url']

In [8]:
((loan_dataset['loan_amnt'] == loan_dataset['funded_amnt']).sum(), 
 (loan_dataset['loan_amnt'] == loan_dataset['funded_amnt']).sum() / len(loan_dataset))

(np.int64(2258603), np.float64(0.9990865531780871))

In [9]:
((loan_dataset['funded_amnt_inv'] == loan_dataset['funded_amnt']).sum(),
(loan_dataset['funded_amnt_inv'] == loan_dataset['funded_amnt']).sum() / len(loan_dataset))

(np.int64(2109556), np.float64(0.9331560406039277))

In [10]:
((loan_dataset['loan_amnt'] == loan_dataset['funded_amnt_inv']).sum(), 
(loan_dataset['loan_amnt'] == loan_dataset['funded_amnt_inv']).sum() / len(loan_dataset))

(np.int64(2109166), np.float64(0.932983525223518))

### 💡 Observation on `loan_amnt`, `funded_amnt`, and `funded_amnt_inv`:

- `loan_amnt` represents the **amount requested** by the borrower.
- `funded_amnt` represents the **amount actually funded** by the bank.
- `funded_amnt_inv` represents the **total amount committed by investors** for that loan at that point in time.

> As per the data,
>
> > **99.91% of the time** the `funded_amnt` was equal to the `loan_amnt`.
> >
> > **93.31% of the time** the `funded_amnt` was equal to the `funded_amnt_inv`.
> >
> > **93.29% of the time** the `funded_amnt_inv` was equal to the `loan_amnt`.
> >

👉 This shows that these three variables are highly correlated with each other and including all three variables would be redundant. We can safely **retain only one** (preferably `funded_amnt`) and drop the other two.

In [11]:
# This assigns a unique ID to each record, but it will be similar to serial number of the df. Hence, not keep it.
# loan_dataset['id'] = range(len(loan_dataset)) 

loan_dataset.drop(columns=['id', 'member_id', 'url', 'loan_amnt', 'funded_amnt_inv'], inplace=True) 

# Dropping id, member id, and URL as all were completely empty.
# Dropping 'loan_amnt', as it have extremely high correlation with 'funded_amnt'

In [12]:
loan_dataset.info() # To get additional information about a dataset.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260668 entries, 0 to 2260667
Columns: 140 entries, funded_amnt to settlement_term
dtypes: float64(101), int64(3), object(36)
memory usage: 2.4+ GB


In [13]:
# To save the current version of our dataset.

loan_dataset.to_csv('dataset/loan1.csv', index=False) 